File: porting/tauri/rust-core/src/image_edit.rs
/// Image editing operations — resample, color adjust, split.
///
/// Replaces `uimageedit.pas`: `ResampleIntfImage` (box filter),
/// `AdjustColors` (fixed per-pixel pipeline), `SplitIntfImage` (parallel cut lines).
///
/// Uses the `image` crate for pixel-level manipulation instead of LCL's TLazIntfImage.

use image::DynamicImage;

// ---------------------------------------------------------------------------
// Resample — box filter resize
// ---------------------------------------------------------------------------

/// Resample an image to new dimensions using a box filter (nearest-neighbor quality).
/// Works for both enlargement and reduction.
///
/// Returns None for invalid input or output dimensions.
pub fn resample_image(img: &DynamicImage, new_w: u32, new_h: u32) -> Option<DynamicImage> {
    if img.width() == 0 || img.height() == 0 || new_w == 0 || new_h == 0 {
        return None;
    }

    // Use the image crate's resize_exact with Triangle filter.
    let resized = image::imageops::resize(img, new_w, new_h, image::imageops::FilterType::Triangle);
    Some(DynamicImage::from(resized))
}

// ---------------------------------------------------------------------------
// Color adjustment pipeline
// ---------------------------------------------------------------------------

/// Color adjustment parameters — mirrors Pascal `TColorAdjust`.
#[derive(Debug, Clone)]
pub struct ColorAdjust {
    pub invert: bool,
    pub grayscale: bool,
    pub sepia: bool,
    pub r_gain: f64,
    pub g_gain: f64,
    pub b_gain: f64,
    pub saturation: f64,
    pub contrast: f64,
    pub brightness: f64,
    pub gamma: f64,
}

impl Default for ColorAdjust {
    fn default() -> Self {
        Self::identity()
    }
}

impl ColorAdjust {
    /// Create an identity (neutral) adjustment.
    pub fn identity() -> Self {
        Self {
            invert: false,
            grayscale: false,
            sepia: false,
            r_gain: 1.0,
            g_gain: 1.0,
            b_gain: 1.0,
            saturation: 1.0,
            contrast: 1.0,
            brightness: 0.0,
            gamma: 1.0,
        }
    }

    /// Check if this adjustment is neutral (no effect).
    pub fn is_neutral(&self) -> bool {
        !self.invert && !self.grayscale && !self.sepia
            && self.r_gain == 1.0 && self.g_gain == 1.0 && self.b_gain == 1.0
            && self.saturation == 1.0 && self.contrast == 1.0
            && self.brightness == 0.0 && self.gamma == 1.0
    }

    /// Apply the adjustment pipeline to a single pixel (R, G, B in 0..255).
    fn transform_pixel(&self, r: f64, g: f64, b: f64) -> (f64, f64, f64) {
        let mut rr = r;
        let mut gg = g;
        let mut bb = b;

        // 1. Invert
        if self.invert {
            rr = 255.0 - rr;
            gg = 255.0 - gg;
            bb = 255.0 - bb;
        }

        // 2. Grayscale (Rec.601 luma)
        if self.grayscale {
            let l = 0.299 * rr + 0.587 * gg + 0.114 * bb;
            rr = l;
            gg = l;
            bb = l;
        }

        // 3. Sepia (standard matrix — uses original R/G/B)
        if self.sepia {
            let orig_r = r;
            let orig_g = g;
            let orig_b = b;
            rr = 0.393 * orig_r + 0.769 * orig_g + 0.189 * orig_b;
            gg = 0.349 * orig_r + 0.686 * orig_g + 0.168 * orig_b;
            bb = 0.272 * orig_r + 0.534 * orig_g + 0.131 * orig_b;
        }

        // 4. Per-channel gains
        rr *= self.r_gain;
        gg *= self.g_gain;
        bb *= self.b_gain;

        // 5. Saturation (mix toward Rec.601 luma)
        if self.saturation != 1.0 {
            let l = 0.299 * rr + 0.587 * gg + 0.114 * bb;
            rr = l + (rr - l) * self.saturation;
            gg = l + (gg - l) * self.saturation;
            bb = l + (bb - l) * self.saturation;
        }

        // 6. Contrast (factor around 128 midpoint)
        if self.contrast != 1.0 {
            rr = (rr - 128.0) * self.contrast + 128.0;
            gg = (gg - 128.0) * self.contrast + 128.0;
            bb = (bb - 128.0) * self.contrast + 128.0;
        }

        // 7. Brightness (additive offset)
        if self.brightness != 0.0 {
            rr += self.brightness;
            gg += self.brightness;
            bb += self.brightness;
        }

        // 8. Gamma
        if self.gamma != 1.0 {
            rr = 255.0 * f64::powf(rr / 255.0, 1.0 / self.gamma);
            gg = 255.0 * f64::powf(gg / 255.0, 1.0 / self.gamma);
            bb = 255.0 * f64::powf(bb / 255.0, 1.0 / self.gamma);
        }

        // Clamp to 0..255
        let rr = rr.max(0.0).min(255.0);
        let gg = gg.max(0.0).min(255.0);
        let bb = bb.max(0.0).min(255.0);

        (rr, gg, bb)
    }
}

/// Apply color adjustments to an image.
///
/// Returns None for invalid input.
pub fn adjust_colors(img: &DynamicImage, adj: &ColorAdjust) -> Option<DynamicImage> {
    if img.width() == 0 || img.height() == 0 {
        return None;
    }

    let mut buf = img.to_rgba8();

    for pixel in buf.pixels_mut() {
        let (r, g, b) = adj.transform_pixel(
            pixel[0] as f64,
            pixel[1] as f64,
            pixel[2] as f64,
        );
        pixel[0] = r.round() as u8;
        pixel[1] = g.round() as u8;
        pixel[2] = b.round() as u8;
        // Alpha preserved from source.
    }

    Some(DynamicImage::from(buf))
}

// ---------------------------------------------------------------------------
// Split
// ---------------------------------------------------------------------------

/// Represents a horizontal or vertical cut line in normalized 0..1 coordinates.
#[derive(Debug, Clone)]
pub struct CutLine {
    pub position: f64, // Normalized position in 0..1
    pub horizontal: bool, // True = horizontal line (cuts vertically)
}

/// Split an image along N parallel cut lines into N+1 pieces.
///
/// Pieces are returned in reading order (top-to-bottom for horizontal lines,
/// left-to-right for vertical lines).
pub fn split_image(img: &DynamicImage, cut_lines: &[CutLine]) -> Option<Vec<DynamicImage>> {
    if img.width() == 0 || img.height() == 0 || cut_lines.is_empty() {
        return None;
    }

    let horizontal = cut_lines[0].horizontal;
    let dim = if horizontal { img.height() } else { img.width() };

    // Sort and normalize cut positions.
    let mut sorted_cuts: Vec<u32> = cut_lines.iter().map(|cl| {
        let pos = (cl.position * dim as f64).round() as u32;
        pos.max(1).min(dim - 1)
    }).collect();

    // Sort ascending and remove duplicates.
    sorted_cuts.sort();
    sorted_cuts.dedup();

    if sorted_cuts.is_empty() {
        return None;
    }

    let mut pieces: Vec<DynamicImage> = Vec::new();

    if horizontal {
        // Horizontal cut lines → vertical strips (top to bottom).
        let mut start_y: u32 = 0;
        for &cut_y in &sorted_cuts {
            let piece_h = cut_y - start_y;
            if piece_h > 0 {
                let cropped = image::imageops::crop_imm(img, 0, start_y, img.width(), piece_h).to_image();
                pieces.push(DynamicImage::from(cropped));
            }
            start_y = cut_y;
        }
        // Last piece.
        let remaining_h = dim - start_y;
        if remaining_h > 0 {
            let cropped = image::imageops::crop_imm(img, 0, start_y, img.width(), remaining_h).to_image();
            pieces.push(DynamicImage::from(cropped));
        }
    } else {
        // Vertical cut lines → horizontal strips (left to right).
        let mut start_x: u32 = 0;
        for &cut_x in &sorted_cuts {
            let piece_w = cut_x - start_x;
            if piece_w > 0 {
                let cropped = image::imageops::crop_imm(img, start_x, 0, piece_w, img.height()).to_image();
                pieces.push(DynamicImage::from(cropped));
            }
            start_x = cut_x;
        }
        // Last piece.
        let remaining_w = dim - start_x;
        if remaining_w > 0 {
            let cropped = image::imageops::crop_imm(img, start_x, 0, remaining_w, img.height()).to_image();
            pieces.push(DynamicImage::from(cropped));
        }
    }

    if pieces.is_empty() {
        None
    } else {
        Some(pieces)
    }
}

File: flutter/app/lib/src/engine/image_edit.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:image/image.dart' as img;

import 'format.dart';

/// Parameters of the colour-adjustment pipeline, applied in this fixed order:
///
/// 1. invert, 2. grayscale, 3. sepia, 4. per-channel gains, 5. saturation,
/// 6. contrast around 128, 7. brightness, 8. gamma (`255*(c/255)^(1/gamma)`).
///
/// Port of `TColorAdjust` from `uimageedit.pas`. All defaults are identity.
class ColorAdjust {
  const ColorAdjust({
    this.invert = false,
    this.grayscale = false,
    this.sepia = false,
    this.rGain = 1.0,
    this.gGain = 1.0,
    this.bGain = 1.0,
    this.saturation = 1.0,
    this.contrast = 1.0,
    this.brightness = 0.0,
    this.gamma = 1.0,
  });

  static const ColorAdjust neutral = ColorAdjust();

  final bool invert;
  final bool grayscale;
  final bool sepia;
  final double rGain;
  final double gGain;
  final double bGain;
  final double saturation;
  final double contrast;
  final double brightness;
  final double gamma;

  bool get isNeutral =>
      !invert &&
      !grayscale &&
      !sepia &&
      rGain == 1.0 &&
      gGain == 1.0 &&
      bGain == 1.0 &&
      saturation == 1.0 &&
      contrast == 1.0 &&
      brightness == 0.0 &&
      gamma == 1.0;
}

/// Box-filter resample (nearest-neighbour quality), both for enlarge and shrink.
/// Port of `ResampleIntfImage` (with its `MaxSamples` stepping). Returns null
/// for invalid input.
img.Image? resampleImage(img.Image source, int newWidth, int newHeight) {
  if (source.width <= 0 || source.height <= 0) return null;
  if (newWidth <= 0 || newHeight <= 0) return null;

  final dst = img.Image(width: newWidth, height: newHeight, numChannels: 4);
  const maxSamples = 4;

  for (var y = 0; y < newHeight; y++) {
    var sy0 = (y * source.height) ~/ newHeight;
    var sy1 = ((y + 1) * source.height) ~/ newHeight;
    if (sy1 <= sy0) sy1 = sy0 + 1;
    final stepY = _step(sy0, sy1, maxSamples);

    for (var x = 0; x < newWidth; x++) {
      var sx0 = (x * source.width) ~/ newWidth;
      var sx1 = ((x + 1) * source.width) ~/ newWidth;
      if (sx1 <= sx0) sx1 = sx0 + 1;
      final stepX = _step(sx0, sx1, maxSamples);

      var r = 0;
      var g = 0;
      var b = 0;
      var a = 0;
      var n = 0;
      for (var iy = sy0; iy < sy1; iy += stepY) {
        for (var ix = sx0; ix < sx1; ix += stepX) {
          final p = source.getPixel(ix, iy);
          r += p.r.toInt();
          g += p.g.toInt();
          b += p.b.toInt();
          a += p.a.toInt();
          n++;
        }
      }
      if (n == 0) continue;
      dst.setPixelRgba(x, y, r ~/ n, g ~/ n, b ~/ n, a ~/ n);
    }
  }
  return dst;
}

int _step(int from, int to, int maxSamples) {
  final span = to - from;
  return span <= 0 ? 1 : ((span + maxSamples - 1) ~/ maxSamples).clamp(1, span);
}

/// Applies [adjust] to [source]. Alpha is preserved. Returns null for invalid
/// input.
img.Image? adjustColors(img.Image source, ColorAdjust adjust) {
  if (source.width <= 0 || source.height <= 0) return null;
  final dst = img.Image(
    width: source.width,
    height: source.height,
    numChannels: 4,
  );
  for (var y = 0; y < source.height; y++) {
    for (var x = 0; x < source.width; x++) {
      final p = source.getPixel(x, y);
      final t = _transform(
        p.r.toDouble(),
        p.g.toDouble(),
        p.b.toDouble(),
        adjust,
      );
      dst.setPixelRgba(x, y, t.$1, t.$2, t.$3, p.a.toInt());
    }
  }
  return dst;
}

(int, int, int) _transform(double r, double g, double b, ColorAdjust adj) {
  if (adj.invert) {
    r = 255 - r;
    g = 255 - g;
    b = 255 - b;
  }
  if (adj.grayscale) {
    final l = 0.299 * r + 0.587 * g + 0.114 * b;
    r = l;
    g = l;
    b = l;
  }
  if (adj.sepia) {
    final sr = 0.393 * r + 0.769 * g + 0.189 * b;
    final sg = 0.349 * r + 0.686 * g + 0.168 * b;
    final sb = 0.272 * r + 0.534 * g + 0.131 * b;
    r = sr;
    g = sg;
    b = sb;
  }
  r *= adj.rGain;
  g *= adj.gGain;
  b *= adj.bGain;
  if (adj.saturation != 1.0) {
    final l = 0.299 * r + 0.587 * g + 0.114 * b;
    r = l + (r - l) * adj.saturation;
    g = l + (g - l) * adj.saturation;
    b = l + (b - l) * adj.saturation;
  }
  if (adj.contrast != 1.0) {
    r = (r - 128) * adj.contrast + 128;
    g = (g - 128) * adj.contrast + 128;
    b = (b - 128) * adj.contrast + 128;
  }
  if (adj.brightness != 0.0) {
    r += adj.brightness;
    g += adj.brightness;
    b += adj.brightness;
  }
  if (adj.gamma != 1.0) {
    r = (255 * math.pow(r / 255, 1 / adj.gamma)).toDouble();
    g = (255 * math.pow(g / 255, 1 / adj.gamma)).toDouble();
    b = (255 * math.pow(b / 255, 1 / adj.gamma)).toDouble();
  }
  return (_clamp255(r), _clamp255(g), _clamp255(b));
}

int _clamp255(double v) => v < 0
    ? 0
    : v > 255
    ? 255
    : v.round();

/// Encodes [image] for [targetExt] using the reference writers: JPEG q92,
/// PNG/BMP lossless, WebP q75; anything else falls back to PNG. [targetExt] is
/// expected to come from [encodeExtFor].
Uint8List encodeImage(img.Image image, String targetExt) {
  switch (targetExt.toLowerCase()) {
    case '.jpg':
    case '.jpeg':
      return Uint8List.fromList(img.encodeJpg(image, quality: 92));
    case '.bmp':
      return Uint8List.fromList(img.encodeBmp(image));
    case '.webp':
      return Uint8List.fromList(
        img.encodeWebP(image, lossless: false, quality: 75),
      );
    case '.png':
    default:
      return Uint8List.fromList(img.encodePng(image));
  }
}

/// Applies an edit pipeline to [src] and returns the encoded page piece(s):
/// optional resize → colour adjust → optional split → encode for [targetExt].
///
/// When [split] is set, the cut positions come from [cuts] (fractions 0..1) when
/// non-empty, otherwise [pieces] equal slices are produced.
List<Uint8List> applyEditPipeline(
  img.Image src, {
  int? width,
  int? height,
  ColorAdjust adjust = ColorAdjust.neutral,
  bool split = false,
  bool horizontal = true,
  int pieces = 2,
  List<double>? cuts,
  required String targetExt,
}) {
  var image = src;
  if (width != null && height != null) {
    image = resampleImage(image, width, height) ?? image;
  }
  if (!adjust.isNeutral) {
    image = adjustColors(image, adjust) ?? image;
  }

  final List<img.Image> images;
  if (split) {
    final effective = (cuts != null && cuts.isNotEmpty)
        ? cuts
        : <double>[for (var i = 1; i < pieces; i++) i / pieces];
    images = splitImage(image, horizontal: horizontal, cuts: effective);
  } else {
    images = [image];
  }
  return [for (final im in images) encodeImage(im, targetExt)];
}

/// Cuts [source] along parallel lines into N+1 pieces, in reading order.
///
/// [cuts] are fractions 0..1; they are clamped to [1, dim-1] pixels, sorted and
/// deduplicated. Returns an empty list when there is no usable cut. Port of
/// `SplitIntfImage`.
List<img.Image> splitImage(
  img.Image source, {
  required bool horizontal,
  required List<double> cuts,
}) {
  if (source.width <= 0 || source.height <= 0 || cuts.isEmpty) {
    return const <img.Image>[];
  }
  final dim = horizontal ? source.height : source.width;
  final normalized = _normalizeCuts(cuts, dim);
  if (normalized.isEmpty) return const <img.Image>[];

  final pieces = <img.Image>[];
  var start = 0;
  for (var i = 0; i <= normalized.length; i++) {
    final end = i < normalized.length ? normalized[i] : dim;
    final size = end - start;
    if (size > 0) {
      pieces.add(
        horizontal
            ? _copyHorizontal(source, start, size)
            : _copyVertical(source, start, size),
      );
    }
    start = end;
  }
  return pieces;
}

List<int> _normalizeCuts(List<double> cutPos, int dim) {
  final cuts = <int>[];
  for (final c in cutPos) {
    var p = (c * dim).round();
    if (p < 1) p = 1;
    if (p > dim - 1) p = dim - 1;
    cuts.add(p);
  }
  cuts.sort();
  final out = <int>[];
  for (final c in cuts) {
    if (out.isNotEmpty && out.last == c) continue;
    out.add(c);
  }
  return out;
}

img.Image _copyHorizontal(img.Image source, int y0, int height) {
  final piece = img.Image(width: source.width, height: height, numChannels: 4);
  for (var y = 0; y < height; y++) {
    for (var x = 0; x < source.width; x++) {
      final p = source.getPixel(x, y0 + y);
      piece.setPixelRgba(
        x,
        y,
        p.r.toInt(),
        p.g.toInt(),
        p.b.toInt(),
        p.a.toInt(),
      );
    }
  }
  return piece;
}

img.Image _copyVertical(img.Image source, int x0, int width) {
  final piece = img.Image(width: width, height: source.height, numChannels: 4);
  for (var y = 0; y < source.height; y++) {
    for (var x = 0; x < width; x++) {
      final p = source.getPixel(x0 + x, y);
      piece.setPixelRgba(
        x,
        y,
        p.r.toInt(),
        p.g.toInt(),
        p.b.toInt(),
        p.a.toInt(),
      );
    }
  }
  return piece;
}
